File: skills/optimize-java-code/assets/examples/comparisons/Pairs.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

/** Java 8-compatible core. Neither variant mutates caller-owned inputs. */
public final class Pairs {
    private Pairs() { }

    public static String baselineJoin(List<String> values) {
        String result = "";
        for (String value : values) result = result + value;
        return result;
    }
    public static String candidateJoin(List<String> values) {
        StringBuilder result = new StringBuilder();
        for (String value : values) result.append(value);
        return result.toString();
    }
    public static Map<String, Integer> baselineCounts(List<String> values) {
        List<String> keys = new ArrayList<>();
        List<Integer> counts = new ArrayList<>();
        for (String value : values) {
            int index = keys.indexOf(value);
            if (index < 0) { keys.add(value); counts.add(1); }
            else counts.set(index, counts.get(index) + 1);
        }
        Map<String, Integer> result = new LinkedHashMap<>();
        for (int i = 0; i < keys.size(); i++) result.put(keys.get(i), counts.get(i));
        return result;
    }
    public static Map<String, Integer> candidateCounts(List<String> values) {
        Map<String, Integer> result = new LinkedHashMap<>();
        for (String value : values) result.merge(value, 1, Integer::sum);
        return result;
    }
    public static List<Boolean> baselineMembership(List<String> values, List<String> queries) {
        List<Boolean> result = new ArrayList<>();
        for (String query : queries) result.add(values.contains(query));
        return result;
    }
    public static List<Boolean> candidateMembership(List<String> values, List<String> queries) {
        Set<String> members = new HashSet<>(values);
        List<Boolean> result = new ArrayList<>();
        for (String query : queries) result.add(members.contains(query));
        return result;
    }
    public static long baselineSum(int[] values) {
        List<Long> temporary = new ArrayList<>();
        for (int value : values) if (value % 2 == 0)
            temporary.add(Math.multiplyExact((long) value, value));
        long result = 0;
        for (Long value : temporary) result = Math.addExact(result, value);
        return result;
    }
    public static long candidateSum(int[] values) {
        long result = 0;
        for (int value : values) if (value % 2 == 0)
            result = Math.addExact(result, Math.multiplyExact((long) value, value));
        return result;
    }
    public static List<Integer> baselineQueue(List<Integer> values) {
        List<Integer> pending = new ArrayList<>(values);
        List<Integer> result = new ArrayList<>();
        while (!pending.isEmpty()) result.add(pending.remove(0));
        return result;
    }
    public static List<Integer> candidateQueue(List<Integer> values) {
        ArrayDeque<Integer> pending = new ArrayDeque<>(values);
        List<Integer> result = new ArrayList<>();
        while (!pending.isEmpty()) result.add(pending.removeFirst());
        return result;
    }
    public static int baselineDelimiters(String text) {
        return text.split(":", -1).length - 1;
    }
    public static int candidateDelimiters(String text) {
        int result = 0;
        for (int i = 0; i < text.length(); i++) if (text.charAt(i) == ':') result++;
        return result;
    }
    public static List<Integer> baselineDistinct(List<Integer> values) {
        List<Integer> result = new ArrayList<Integer>();
        for (Integer value : values) if (!result.contains(value)) result.add(value);
        return result;
    }
    public static List<Integer> candidateDistinct(List<Integer> values) {
        Set<Integer> seen = new HashSet<Integer>();
        List<Integer> result = new ArrayList<Integer>();
        for (Integer value : values) if (seen.add(value)) result.add(value);
        return result;
    }
    public static List<Integer> baselineReverse(List<Integer> values) {
        List<Integer> result = new ArrayList<Integer>();
        for (Integer value : values) result.add(0, value);
        return result;
    }
    public static List<Integer> candidateReverse(List<Integer> values) {
        List<Integer> result = new ArrayList<Integer>(values);
        java.util.Collections.reverse(result);
        return result;
    }
    public static List<Long> baselinePositiveSquares(int[] values) {
        return Arrays.stream(values)
            .filter(value -> value > 0)
            .mapToLong(value -> (long) value * value)
            .boxed()
            .collect(Collectors.toList());
    }
    public static List<Long> candidatePositiveSquares(int[] values) {
        List<Long> result = new ArrayList<Long>(values.length);
        for (int value : values) if (value > 0) result.add((long) value * value);
        return result;
    }
    private static void equal(Object a, Object b) {
        if (!a.equals(b)) throw new IllegalStateException(a + " != " + b);
    }
    public static void verify() {
        int checks = 0;
        for (int length = 0; length <= 5; length++) {
            int possibilities = (int) Math.pow(3, length);
            for (int code = 0; code < possibilities; code++) {
                int rest = code;
                int[] numbers = new int[length];
                List<Integer> integers = new ArrayList<>();
                List<String> strings = new ArrayList<>();
                for (int i = 0; i < length; i++) {
                    numbers[i] = rest % 3 - 1; rest /= 3;
                    integers.add(numbers[i]); strings.add(Integer.toString(numbers[i]));
                }
                int[] saved = numbers.clone();
                equal(baselineJoin(strings), candidateJoin(strings));
                equal(new ArrayList<>(baselineCounts(strings).entrySet()), new ArrayList<>(candidateCounts(strings).entrySet()));
                equal(baselineMembership(strings, Arrays.asList("0", "9")), candidateMembership(strings, Arrays.asList("0", "9")));
                equal(baselineSum(numbers), candidateSum(numbers));
                equal(baselineQueue(integers), candidateQueue(integers));
                equal(baselineDelimiters(String.join(":", strings)), candidateDelimiters(String.join(":", strings)));
                equal(baselinePositiveSquares(numbers), candidatePositiveSquares(numbers));
                if (!Arrays.equals(saved, numbers)) throw new IllegalStateException("mutated input");
                checks += 8;
            }
        }
        List<String> unicode = Arrays.asList("é", "e\u0301", "é", "🙂", "\0");
        Map<String, Integer> expected = new LinkedHashMap<>();
        expected.put("é", 2); expected.put("e\u0301", 1); expected.put("🙂", 1); expected.put("\0", 1);
        equal(baselineCounts(unicode), expected); equal(candidateCounts(unicode), expected);
        equal(baselineJoin(Arrays.asList("a", "", "é", "🙂")), "aé🙂");
        equal(candidateJoin(Arrays.asList("a", "", "é", "🙂")), "aé🙂");
        equal(baselineDelimiters(":é::🙂:"), 4); equal(candidateDelimiters(":é::🙂:"), 4);
        equal(baselineDelimiters(""), 0); equal(candidateDelimiters(""), 0);
        equal(baselineSum(new int[] {-3, 2, 2, 0, 5}), 8L);
        equal(candidateSum(new int[] {-3, 2, 2, 0, 5}), 8L);
        equal(baselineQueue(Arrays.asList(2, 1, 2)), Arrays.asList(2, 1, 2));
        equal(candidateQueue(Arrays.asList(2, 1, 2)), Arrays.asList(2, 1, 2));
        for (boolean candidate : new boolean[] {false, true}) {
            boolean threw = false;
            try {
                int[] overflow = {Integer.MIN_VALUE, Integer.MIN_VALUE};
                if (candidate) candidateSum(overflow); else baselineSum(overflow);
            } catch (ArithmeticException expectedOverflow) { threw = true; }
            if (!threw) throw new IllegalStateException("overflow contract lost");
        }
        // A delimiter-preserving split is essential. Default split drops trailing empties.
        equal("a:".split(":").length, 1);
        equal("a:".split(":", -1).length, 2);
        System.out.println("PASS differential and expected result checks: " + checks + " + edge cases");
    }
    public static void main(String[] args) {
        if (args.length == 1 && args[0].equals("verify")) { verify(); return; }
        if (args.length != 3 || !(args[0].equals("baseline") || args[0].equals("candidate")))
            throw new IllegalArgumentException("usage: Pairs verify | baseline|candidate CASE SIZE");
        int which = Integer.parseInt(args[1]), size = Integer.parseInt(args[2]);
        if (size < 0 || size > 100000) throw new IllegalArgumentException("size 0..100000");
        boolean candidate = args[0].equals("candidate");
        int[] numbers = new int[size];
        List<Integer> integers = new ArrayList<>();
        List<String> strings = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            numbers[i] = i % 31 - 15; integers.add(numbers[i]);
            strings.add(Integer.toString(numbers[i]));
        }
        Object result;
        switch (which) {
            case 1: result = candidate ? candidateJoin(strings) : baselineJoin(strings); break;
            case 2: result = candidate ? candidateCounts(strings) : baselineCounts(strings); break;
            case 3: result = candidate ? candidateMembership(strings, strings) : baselineMembership(strings, strings); break;
            case 4: result = candidate ? candidateSum(numbers) : baselineSum(numbers); break;
            case 5: result = candidate ? candidateQueue(integers) : baselineQueue(integers); break;
            case 6: String text = String.join(":", strings);
                    result = candidate ? candidateDelimiters(text) : baselineDelimiters(text); break;
            case 7: result = candidate ? candidateSum(numbers) : baselineSum(numbers); break;
            case 8: result = candidate ? candidatePositiveSquares(numbers) : baselinePositiveSquares(numbers); break;
            default: throw new IllegalArgumentException("case 1..8");
        }
        System.out.println(result);
    }
}

File: skills/optimize-csharp-code/assets/examples/comparisons/Pairs.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;

public static class Pairs
{
    public static string BaselineJoin(string[] values)
    {
        string result = "";
        foreach (string value in values)
            result += value;
        return result;
    }

    public static string CandidateJoin(string[] values)
    {
        var result = new StringBuilder();
        foreach (string value in values)
            result.Append(value);
        return result.ToString();
    }

    public static KeyValuePair<string, int>[] BaselineCounts(string[] values)
    {
        var result = new List<KeyValuePair<string, int>>();
        foreach (string value in values)
        {
            int index = result.FindIndex(pair => StringComparer.Ordinal.Equals(pair.Key, value));
            if (index < 0)
                result.Add(new(value, 1));
            else
                result[index] = new(value, result[index].Value + 1);
        }
        return result.ToArray();
    }

    public static KeyValuePair<string, int>[] CandidateCounts(string[] values)
    {
        var indices = new Dictionary<string, int>(StringComparer.Ordinal);
        var result = new List<KeyValuePair<string, int>>();
        foreach (string value in values)
        {
            if (indices.TryGetValue(value, out int index))
                result[index] = new(value, result[index].Value + 1);
            else
            {
                indices.Add(value, result.Count);
                result.Add(new(value, 1));
            }
        }
        return result.ToArray();
    }

    public static bool[] BaselineMembership(string[] values, string[] queries) =>
        queries
            .Select(query =>
                Array.Exists(values, value => StringComparer.Ordinal.Equals(value, query))
            )
            .ToArray();

    public static bool[] CandidateMembership(string[] values, string[] queries)
    {
        var members = new HashSet<string>(values, StringComparer.Ordinal);
        return queries.Select(members.Contains).ToArray();
    }

    public static long BaselineSum(int[] values) =>
        values
            .Where(value => value % 2 == 0)
            .Select(value => checked((long)value * value))
            .ToArray()
            .Sum();

    public static long CandidateSum(ReadOnlySpan<int> values)
    {
        long sum = 0;
        foreach (int value in values)
            if (value % 2 == 0)
                sum = checked(sum + checked((long)value * value));
        return sum;
    }

    public static int BaselineParse(string text, int start, int count) =>
        int.Parse(text.Substring(start, count), NumberStyles.Integer, CultureInfo.InvariantCulture);

    public static int CandidateParse(string text, int start, int count) =>
        int.Parse(text.AsSpan(start, count), NumberStyles.Integer, CultureInfo.InvariantCulture);

    public static int[] BaselineQueue(int[] values)
    {
        var pending = new List<int>(values);
        var result = new List<int>();
        while (pending.Count != 0)
        {
            result.Add(pending[0]);
            pending.RemoveAt(0);
        }
        return result.ToArray();
    }

    public static int[] CandidateQueue(int[] values)
    {
        var pending = new Queue<int>(values);
        var result = new List<int>();
        while (pending.TryDequeue(out int value))
            result.Add(value);
        return result.ToArray();
    }

    public static int[] BaselineDistinct(int[] values)
    {
        var result = new List<int>();
        foreach (int value in values)
            if (!result.Contains(value)) result.Add(value);
        return result.ToArray();
    }

    public static int[] CandidateDistinct(int[] values)
    {
        var seen = new HashSet<int>();
        var result = new List<int>();
        foreach (int value in values)
            if (seen.Add(value)) result.Add(value);
        return result.ToArray();
    }

    public static int BaselineDelimiterCount(string value) => value.Split(':', StringSplitOptions.None).Length - 1;
    public static int CandidateDelimiterCount(string value) => value.Count(character => character == ':');
    public static int[] BaselineReverse(int[] values) => values.Aggregate(new List<int>(), (result, value) => { result.Insert(0, value); return result; }).ToArray();
    public static int[] CandidateReverse(int[] values) => values.Reverse().ToArray();
    public static long[] BaselinePositiveSquares(int[] values) => values.Where(value => value > 0).Select(value => (long)value * value).ToArray();
    public static long[] CandidatePositiveSquares(ReadOnlySpan<int> values)
    {
        var result = new List<long>();
        foreach (int value in values) if (value > 0) result.Add((long)value * value);
        return result.ToArray();
    }

    public static long BaselineBoxedSum(int[] values)
    {
        object[] boxed = values.Cast<object>().ToArray();
        long result = 0;
        foreach (object value in boxed)
            result = checked(result + (int)value);
        return result;
    }

    public static long CandidateSpanSum(ReadOnlySpan<int> values)
    {
        long result = 0;
        foreach (int value in values)
            result = checked(result + value);
        return result;
    }
}
